### ISSUE TITLE
[Question] Make predicitions about derivatives (mean *and* variance)

### ISSUE THREAD

**Issue by trysetnull** [cornellius-gp/gpytorch#1116]
The example: `gpytorch/examples/08_Advanced_Usage/Simple_GP_Regression_Derivative_Information_1d.ipynb`
shows us how to make predictions about derivatives, and also to **make inference based on derivative information**.

What is the best way to go about making predictions about derivatives when derivative information isn't available? In other words I would like to be able to not only draw samples from the trained GP but also the first derivatives.

I tried passing `gpytorch.distributions.MultivariateNormal` the first gradient block (`K[..., :n1, n2:]`) returned from `gpytorch.kernels.RBFKernelGrad` but without success.

**Comment by KeAWang** [comment]
Is #772 what you're looking for?

**Comment by trysetnull** [comment]
Thanks for the rapid response!
Yes, the follow shows exactly what I'm looking for:
https://github.com/cornellius-gp/gpytorch/issues/772#issuecomment-508124687

**Comment by KeAWang** [comment]
No problem! You probably want to use the next comment (with the sum) which is much faster

**Comment by trysetnull** [comment]
With your help, I now know how to calculate the first derivative of the mean function.

Next I turned my attention to calculating the first derivative from a drawn sample, I did it like this:
```
sample = f_preds.rsample(sample_shape=torch.Size((1,)))
dsample_dtest_x = torch.tensor([torch.autograd.grad(sample, test_x, retain_graph=True)[0][i] for i in range(test_x.shape[0])])
```

Unfortunately, I haven't managed to calculate the **_variance_** of the first derivative.
In principal I know how to do it:

`V|f_*'| = K_X*X*'' - K_X*X' (K_XX + sigma^2 I)^-1 K_X*X'`

i.e. the k_** matrix (test_test) needs to be calculated to second order and the k_* terms (test_train) to first order.

I've taken a look at [gpytorch/models/exact_prediction_strategies.py](https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/models/exact_prediction_strategies.py) but I'm still none the wiser.

:see_no_evil: 

I'd really appreciate any hints!

**Comment by KeAWang** [comment]
There's no an easy way of computing the variance analytically from the trained `model`. But you can get an empirical estimate from sampling.

Modifying the same code as https://github.com/cornellius-gp/gpytorch/issues/772#issuecomment-508124687:

```python
# Get into evaluation (predictive posterior) mode
model.eval()
likelihood.eval()

# Test points are regularly spaced along [0,1]
# Make predictions by feeding model through likelihood
with gpytorch.settings.fast_pred_var():
    test_x = torch.linspace(0, 1, 51, requires_grad=True)
    observed_pred = likelihood(model(test_x))
    dydtest_x = torch.autograd.grad(observed_pred.mean.sum(), test_x, retain_graph=True)[0]

    n_samples = 100
    sampled_pred = observed_pred.rsample(torch.Size([n_samples]))
    sampled_dydtest_x = torch.stack([torch.autograd.grad(pred.sum(), test_x, retain_graph=True)[0] for pred in sampled_pred])
    dydtest_x_sample_std = sampled_dydtest_x.std(0)
    lower = dydtest_x - 2 * dydtest_x_sample_std
    upper = dydtest_x + 2 * dydtest_x_sample_std
    

with torch.no_grad():
    # Initialize plot
    f, ax = plt.subplots(1, 1, figsize=(4, 3))

    # Plot training data as black stars
    ax.plot(train_x.numpy(), train_y.numpy(), 'k*')
    # Plot predictive means as blue line
    ax.plot(test_x.detach().numpy(), observed_pred.mean.detach().numpy(), 'b')
    # Plot expected derivative
    ax.plot(test_x.detach().numpy(), dydtest_x.numpy(), 'r')
    # Plot 2 times empirical standard deviations about expected derivative
    ax.fill_between(test_x.detach().numpy(), lower.detach().numpy(), upper.detach().numpy(), alpha=0.1, color="r")
    # Plot real derivative
    ax.plot(test_x.detach().numpy(), 2 * math.pi * torch.cos(2 * math.pi * test_x).detach().numpy(), 'g')
    ax.set_ylim([-7, 7])
    ax.legend(['Observed Data', 'Mean',  'Estimated Derivative', 'True Derivative'])
```

This gives 
![image](https://user-images.githubusercontent.com/11478740/79679232-1bef3c00-81d2-11ea-86ab-97dd3cc60279.png)
